### ISSUE TITLE
Rustc autocompletions assume rustc is in the $PATH (uses wrong rustc)

### ISSUE THREAD

**Issue by nagisa** [fish-shell/fish-shell#3117]
I have no `rustc` in my `$PATH`.

Typing `~/.cargo/bin/rustc t` spews following:

```
/usr/share/fish/completions/rustc.fish (line 1): rustc -C help \
                                                 ^
in command substitution
    called on line -1 of file /usr/share/fish/completions/rustc.fish

from sourcing file /usr/share/fish/completions/rustc.fish
    called on standard input

in command substitution
    called on standard input

fish: Unknown command 'rustc'
/usr/share/fish/completions/rustc.fish (line 1): rustc -Z help \
                                                 ^
in command substitution
    called on line -1 of file /usr/share/fish/completions/rustc.fish

from sourcing file /usr/share/fish/completions/rustc.fish
    called on standard input

in command substitution
    called on standard input

fish: Unknown command 'rustc'
/usr/share/fish/completions/rustc.fish (line 1): rustc -W help  \
                                                 ^
in command substitution
    called on line -1 of file /usr/share/fish/completions/rustc.fish

from sourcing file /usr/share/fish/completions/rustc.fish
    called on standard input

in command substitution
    called on standard input
```
### Reproduction Steps:
1. Install rust through rustup/have no rust in $PATH
2. type `./$PATH_TO_RUSTC something`
3. observe error
### Expected behavior:

the autocompletion service should use rustc at the full path that’s being autocompleted
### Observed behavior:

Instead `rustc` is looked for in `$PATH`
### Additional information:

fish, version 2.3.0, arch linux, rustc via rustup.

**Comment by faho** [comment]
This is a much more general problem - it would also be triggered by doing the same thing with e.g. git. Any completion that calls the tool to be completed. It also touches on #2365.

Adding the behavior you suggested is going to be rather tricky - we'd need to fix every completion that does this, and we'd need to be really careful, because it's going to look like this:

```fish
# Instead of
complete -c rustc -a '(rustc --something)'
# do
complete -c rustc -a '(eval (commandline -opc)[-1] (string escape -- --something))'
```

We could also fake the completion environment to have the path to the completed tool be in the PATH - but that would break if there's more stuff in there that you don't want in your PATH (which is presumably why you haven't added it).

Or we could just fake the call to the completed tool to point to that path, but that's not really all that elegant and might also have weird corner-cases.

What's your reason for not adding this (or a symlink to it) to $PATH?

**Comment by nagisa** [comment]
> What's your reason for not adding this (or a symlink to it) to $PATH?

I use the rustc I have installed on my system rarely enough to justify the effort of adding it to the PATH. Being a developer of rustc, I also happen to have multiple rustc binaries on my system all around the place, which would make fish give autocompletions from the wrong rustc most of the time anyway.

**Comment by faho** [comment]
As a workaround, try this:

```fish
function rustc --description 'Hack to make completions use the specified rustc'
    set -l cmd (commandline -opc)
    if string match -q "*rustc" -- $cmd[-1]
        # Because we run eval, we also need to escape all arguments to it.
        eval (string escape -- $cmd[-1] $argv)
    else
        # We haven't been given a rust, so just try calling the normal one.
        command rustc $argv
    end
end
```

You can also make a copy of the rustc completions for yourself and call that function (which you then should rename) instead of `rustc`, but note that the completions generate the options once, so if you switch rustcs, you need to re-source them or start a new shell.

I'm not adding that hack here, and the real solution might take a while.

**Comment by zanchey** [comment]
I wonder if completions should define a `COMPLETE_CMD` function that calls the command being completed.

**Comment by mqudsi** [comment]
I believe this is closed by 4414d5c8885b5cd53c7ab42f70bd2824050907f0. Please comment if reopening is needed.

**Comment by faho** [comment]
Nope, not quite closed. This is about the user calling something like `~/bin/rustc`, and the completion script then going and using `/usr/bin/rustc` (via $PATH).

While it does fix the case where /usr/bin/rustc doesn't exist, the real issue is that ~/bin/rustc should be called. We haven't quite figured out how, though.

**Comment by mqudsi** [comment]
Thanks, I updated the title. 

Either we do a mass update with `COMPLETE_CMD` or whatever as @zanchey had suggested or we hack around it:

```fish
~> echo $PATH
/usr/local/bin:/usr/bin
~> /foo/bar/baz <TAB>
```

This would would trigger `baz.fish` for the completions, which then calls `baz` to generate the completions, but calls the wrong one.

But if in `complete.cpp` before calling `baz.fish` we did something like 

```cpp
const wcstring binary_dir = resolve_dir(cmd);
env_block_t env2 = env.clone();
env2.set("PATH", binary_dir.append(":").append(env2.get("PATH")));
std::swap(env2, env);
// call complete_autoload(....)
std::swap(env2, env);
// continue
```

then when `baz.fish` executes `baz`, the `baz` that the user typed at the prompt would be used.

But that's a big hack for something like this. How often do the completions for the copy that's in the PATH not line up with the copy the user is using? Might be better to just not offer suggestions if an absolute path to a binary is used and that absolute path's parent dir is not in PATH.
